### DISCUSSION TITLE
Custom Login form with OTP configuration

### DISCUSSION THREAD

**Discussion by ashanjayasundara** [keycloak/keycloak#20768] [Q&A]
I am trying to implement customize the out flow and login form which show users to custom form which having configure OTP now or skip for now (n days until this is required) options once their initial login success. Let me know if it is addressable within current features
something similar to this.[Img](https://www.wintips.org/wp-content/uploads/2022/12/image_thumb-141.png)

**Comment by initshdb** [comment]
One way to achieve the similar is overriding the existing OTP Form - its Freemarker template, Required Action Provider and Authenticator, as follows.

- Extend the Keycloak login theme and override the `login-config-totp.ftl`. This is shown when the `OTP Form` is included to the Authentication Flow. If the OTP authenticator is not configured for the user then after the username password login prompt the user is required to shown the OTP form to configure the OTP authenticator.
- In the `login-config-totp.ftl` add a conditional button below the `Submit` button which will show a message like `Skip for now ('X' days until this is required`.
- Extend the `UpdateTotp.java` Required Action Provider as your own `CustomUpdateTotp.java` Required Action Provider. Here in the `requiredActionChallenge()` you can to add 2 attributes - `skipForNowUntilDays`: no. of days based on your logic until to mandate configuring the OTP authenticator, and a boolean `forceConfigureOTP` which will conditionally show or hide the 'Skip' button based on the no. of days > or < 0. If the 'Skip' button is shown the user should be able to login in without configuring the OTP, and if the button is hidden then the user is forced o configure the OTP.
- You need to create a `CustomOTPFormAuthenticator` and `CustomOTPFormAuthenticatorFactory` copying or extending the original `OTPFormAuthenticator` and `OTPFormAuthenticatorFactory`. Replace the required action in `setRequiredActions` and `getRequiredActions` to the `CustomUpdateTotp` instead of the original `UpdateTotp`.
- Finally configure this `Custom OTP Form` authenticator instead of the `OTP Form` in the Authentication flow. The extended [OTP Form should be shown](https://github.com/keycloak/keycloak/assets/19492841/be7fdc9e-e44e-4c2b-9278-75bfb8569909). On click of `Skip` button, you need to call `context.success()` in the `processAction()` of `CustomUpdateTotp`.
